Update design overview to reflect submitted proposals (#3180)
carbon-language/carbon-lang: docs/design/README.md

-   Function call syntax is used to specify parameters to a type, like
    `HashMap(String, i64)`.

> References:
>
> -   Proposal
>     [#2360: Types are values of type `type`](https://github.com/carbon-language/carbon-lang/pull/2360)

### Values usable as types

A value used in a type position, like after a `:` in a variable declaration or
the return type after a `->` in a function declaration, must be:

-   [a compile-time constant](#expression-phases), so the compiler can evaluate
    it at compile time, and

> -   have a defined implicit conversion to type `type`.

Reviewer: ```suggestion
the return type after a `->` in a function declaration, must:

-   be [a compile-time constant](#expression-phases), so the compiler can evaluate
    it at compile time, and
-   have a defined implicit conversion to type `type`.
```

(Avoid "must be [...] have a" for the second bullet.)

One difference from C++ templates, Carbon template instantiation is not
controlled by the SFINAE rule of C++
([1](https://en.wikipedia.org/wiki/Substitution_failure_is_not_an_error),
[2](https://en.cppreference.com/w/cpp/language/sfinae)) but by explicit
constraints declared in the function signature and evaluated at compile-time.

> **TODO:** The design for template constraints is still under development.

The [expression phase](#expression-phases) of a checked parameter is a symbolic
constant whereas the expression phase of a template parameter is template
constant. A binding pattern using `:!` is a _compile-time binding pattern_; more
specifically a _template binding pattern_ if it uses `template`, and a _symbolic
binding pattern_ if it does not.

Although checked generics are generally preferred, templates enable translation
of code between C++ and Carbon, and address some cases where the type checking

> rigor of checked generics are problematic.

Reviewer: ```suggestion
rigor of checked generics is problematic.
```
(Minor grammar fix since we're in the area.)

> -   Proposal
>     [#624: Coherence: terminology, rationale, alternatives considered](https://github.com/carbon-language/carbon-lang/pull/624)
> -   Proposal
>     [#989: Member access expressions](https://github.com/carbon-language/carbon-lang/pull/989)
> -   Proposal
>     [#990: Generics details 8: interface default and final members](https://github.com/carbon-language/carbon-lang/pull/990)
> -   Proposal
>     [#1084: Generics details 9: forward declarations](https://github.com/carbon-language/carbon-lang/pull/1084)
> -   Question-for-leads issue
>     [#1132: How do we match forward declarations with their definitions?](https://github.com/carbon-language/carbon-lang/issues/1132)
> -   Proposal
>     [#2360: Types are values of type `type`](https://github.com/carbon-language/carbon-lang/pull/2360)
> -   Proposal
>     [#2760: Consistent `class` and `interface` syntax](https://github.com/carbon-language/carbon-lang/pull/2760)

### Combining constraints

> A function can require calling types to implement multiple interfaces (or other

Reviewer: This is pre-existing, but I had some trouble understanding what a "calling type" is. Would "parameter types" or "type arguments" (or maybe more verbosely, "deduced or explicit type arguments") work here?

Author: Went with "type arguments"

the caller, _in addition_ to any constraints. This means member name lookup and
type checking for anything [dependent](generics/terminology.md#dependent-names)
on the template parameter can't be completed until the template is instantiated
with a specific concrete type. When the constraint is just `type`, this gives
semantics similar to C++ templates.

```carbon
class Game {
  fn Draw[self: Self]() -> bool;
  impl as Renderable {
    fn Draw[self: Self]();
  }
}

fn TemplateDraw[template T:! type](x: T) {
  // Calls `Game.Draw` when `T` is `Game`:
  x.Draw();
}

fn ConstrainedTemplateDraw[template T:! Renderable](x: T) {
  // ❌ Error when `T` is `Game`: Finds both `T.Draw` and
  // `Renderable.Draw`, and they are different.
  x.Draw();
}

fn CheckedGenericDraw[T:! Renderable](x: T) {
  // Always calls `Renderable.Draw`, even when `T` is `Game`:
  x.Draw();
}
```

This allows a safe transition from template to checked generics. Constraints can
be added incrementally, with the compiler verifying that the semantics stay the
same. If adding the constraint would change which function gets called, an error
is triggered, as in `ConstrainedTemplateDraw` from the example. Once all

> constraints have been added, removing the word `template` to switch to a checked
> parameter is safe.

Reviewer: Optional, might make this easier to read:
```suggestion
constraints have been added, it is safe to remove the word `template` to switch
to a checked parameter.
```

    same data representation as an existing type, so you may cast between the
    two types, but can implement different interfaces or implement interfaces
    differently.

> -   Additional requirements can be placed on the associated constants of an
>     interface using
>     [`where` constraints](generics/details.md#where-constraints).

Reviewer: I think this is currently only possible for associated facets, because `where` produces a facet type. Maybe:
```suggestion
-   Additional requirements can be placed on the associated facets of an
    interface using
    [`where` constraints](generics/details.md#where-constraints).
```

> -   [Implied constraints](generics/details.md#implied-constraints) allows some

Reviewer: ```suggestion
-   [Implied constraints](generics/details.md#implied-constraints) allow some
```
Not sure about this: is "implied constraints" a singular noun, short for "the implied constraints feature" or "the use of implied constraints" or similar, or the plural of "implied constraint"?

    constraints to be deduced and omitted from a function signature.
-   _Planned_ [dynamic erased types](generics/details.md#runtime-type-fields)

>     can hold any value with a type implementing an interface, and allows the

Reviewer: ```suggestion
    can hold any value with a type implementing an interface, and allow the
```
